synapse/storage/databases/main/events_worker.py

```diff
@@ -2117,13 +2127,33 @@ async def get_event_id_for_timestamp(
             The closest event_id otherwise None if we can't find any event in
             the given direction.
         """
+        if direction == "b":
+            # Find closest event *before* a given timestamp. We use descending
+            # (which gives values largest to smallest) because we want the
+            # largest possible timestamp *before* the given timestamp.
+            comparison_operator = "<="
+            order = "DESC"
+        else:
+            # Find closest event *after* a given timestamp. We use ascending
+            # (which gives values smallest to largest) because we want the
+            # closest possible timestamp *after* the given timestamp.
+            comparison_operator = ">="
+            order = "ASC"
 
-        sql_template = """
+        sql_template = f"""
             SELECT event_id FROM events
             LEFT JOIN rejections USING (event_id)
             WHERE
-                origin_server_ts %s ?
-                AND room_id = ?
+                room_id = ?
+                AND origin_server_ts {comparison_operator} ?
+                /**
+                 * Make sure the event isn't an `outlier` because we have no way
+                 * to later check whether it's next to a gap. `outliers` do not
+                 * have entries in the `event_edges`, `event_forward_extremeties`,
+                 * and `event_backward_extremities` tables to check against
+                 * (used by `is_event_next_to_backward_gap` and `is_event_next_to_forward_gap`).
+                 */
+                AND outlier = ? /* False */
```

Comment:
What way do we want to do this matching?

I see it done the way we have done it here and this way:

https://github.com/matrix-org/synapse/blob/2c63cdcc3f1aa4625e947de3c23e0a8133c61286/synapse/storage/databases/main/stream.py#L578

Reply:
I can't speak to the relative merits but I do think that `AND NOT outlier` is a little easier to parse.

Reply:
Updated, makes sense to me 👍

Reply:
I think the `AND outlier = ?` style is cargo-culted from `INSERT INTO tbl VALUES(?)`, for which see https://matrix-org.github.io/synapse/latest/development/database_schema.html#boolean-columns.

Mind you, that is out of date now that we have dropped support for old SQLite. `TRUE` and `FALSE` work properly in modern sqlite.

Anyway, TL;DR: `AND NOT outlier` for the win here.